cmd/soroban-rpc/internal/daemon/daemon.go

```diff
@@ -299,28 +300,49 @@ func (d *Daemon) mustInitializeStorage(cfg *config.Config) *feewindow.FeeWindows
 	defer cancelReadTxMeta()
 	var initialSeq uint32
 	var currentSeq uint32
-	dataMigrations, err := db.BuildMigrations(readTxMetaCtx, d.logger, d.db, cfg)
+	applicableRange, err := db.GetMigrationLedgerRange(readTxMetaCtx, d.db, cfg.HistoryRetentionWindow)
 	if err != nil {
-		d.logger.WithError(err).Fatal("could not build migrations")
+		d.logger.WithError(err).Fatal("could not get ledger range for migration")
 	}
 
 	// Merge migrations range and fee stats range to get the applicable range
 	latestLedger, err := db.NewLedgerEntryReader(d.db).GetLatestLedgerSequence(readTxMetaCtx)
-	if err != nil {
+	if err != nil && !errors.Is(err, db.ErrEmptyDB) {
 		d.logger.WithError(err).Fatal("failed to get latest ledger sequence: %w", err)
 	}
-
 	maxFeeRetentionWindow := max(cfg.ClassicFeeStatsLedgerRetentionWindow, cfg.SorobanFeeStatsLedgerRetentionWindow)
-	ledgerSeqRange := &db.LedgerSeqRange{FirstLedgerSeq: latestLedger - maxFeeRetentionWindow, LastLedgerSeq: latestLedger}
-	applicableRange := dataMigrations.ApplicableRange()
+	ledgerSeqRange := &db.LedgerSeqRange{FirstLedgerSeq: firstLedger, LastLedgerSeq: latestLedger}
+	if latestLedger > maxFeeRetentionWindow {
+		ledgerSeqRange.FirstLedgerSeq = latestLedger - maxFeeRetentionWindow
+	}
+
+	// Combine the ledger range for fees, events and transactions
 	ledgerSeqRange = ledgerSeqRange.Merge(applicableRange)
 
+	// Start a common db transaction for the entire migration duration
+	err = d.db.Begin(readTxMetaCtx)
+	if err != nil {
+		d.logger.WithError(err).Fatal("could not commit database transaction: ", d.db.Rollback())
+	}
+	defer func() {
+		err = d.db.Commit()
+		if err != nil {
+			d.logger.WithError(err).Fatal("could not commit database transaction")
+		}
```

Comment:
this will end up the committing the transaction even if `mustInitializeStorage()` panics later on via a fatal log line. So, we need to make sure that the transaction is rolled back before the commit call whenever we intend to exit via an error